File: engine/nexum-core/src/main/java/io/nexum/core/services/HelperSerializerService.java
package io.nexum.core.services;

import io.nexum.core.storages.HelperSerializerRegistry;
import io.nexum.exceptions.HelperSerializationException;
import io.nexum.channel.FriendlyBuffer;
import io.nexum.channel.HelperSerializer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class HelperSerializerService {
    private static @NotNull HelperSerializerService instance = new HelperSerializerService();

    private HelperSerializerService() {

    }

    public <T> void serialize(@NotNull T object, @NotNull FriendlyBuffer friendlyBuffer) {
        @Nullable
        final HelperSerializer<Object> helperSerializer = HelperSerializerRegistry.getInstance().get(object.getClass());

        if(helperSerializer == null) throw new HelperSerializationException(
                String.format("O serializador para o objeto %s não foi encontrado", object.getClass().getSimpleName())
        );

        helperSerializer.serialize(object, friendlyBuffer);
        friendlyBuffer.write("identifier", helperSerializer.getIdentifier());
    }

    public static @NotNull HelperSerializerService getInstance() {
        return instance;
    }
}

File: framework/nexum-core/lib/services/helper_serializer_service.dart
import 'package:nexum_core/exceptions/helper_serialization_exception.dart';

import '../foundation/channel/friendly_buffer.dart';
import '../foundation/channel/helper_serializer.dart';
import '../storages/helper_serializer_registry.dart';

class HelperSerializerService {
  static HelperSerializerService get instance => _instance;
  static const HelperSerializerService _instance = HelperSerializerService._();

  const HelperSerializerService._();

  void serializeObject<T>(T object, FriendlyBuffer friendlyBuffer) {
    final HelperSerializer ? helperSerializer = HelperSerialzerRegistry.instance.getSerializer(object.runtimeType);

    if(helperSerializer == null) {
      throw HelperSerializerException(
        "Não foi possivel encontrar o serializador para o objeto ${object.runtimeType}"
      );
    }

    helperSerializer.serialize(object, friendlyBuffer);
    friendlyBuffer.write("identifier", helperSerializer.identifier);
  }
}